Add the ability to write junit test results file for pytest runs.

asherf opened the pull request:
### Problem

With the transition of pytest to the v2 engine, we lost the ability to write junit-xml results files when test runs.
This can't be solved by passing options thru directly to pytest since it expect a junit-xml result file name, which means we need a unique name for each target.

### Solution

Introduce the ability to calculate the junit-xml file name based on a user provided direcotry (option) and the test target name.

### Result

JUnit xml result files which can be processed by CI (CircleCI for example) so failing test can be seen w/o digging thru endless logs.

https://docs.pytest.org/en/latest/usage.html#creating-junitxml-format-files

asherf commented:
Happy to take feedback and improve this implementation... as I am not a 100% this is the optimal way to address this issue.

asherf commented:
@stuhood @Eric-Arellano

stuhood commented on src/python/pants/backend/python/rules/pytest_runner.py:
```diff
@@ -210,11 +212,18 @@ async def setup_pytest_for_target(
     specified_source_file_names = sorted(specified_source_files.snapshot.files)
+    results_dir = test_options.values.results_dir
+    results_file_path = (
+        (Path(results_dir) / f"{config.address.path_safe_spec}.xml").absolute().as_posix()
```
Absolute paths aren't safe to use from rules... sorry: we need to do a better job of sandboxing so that that is clearer.

Instead, you should use a relative path, which will end up inside the snapshot that is captured by the test run, and then mark the file/directory as captured here: https://github.com/pantsbuild/pants/blob/79b966ce77f746995e357ee06d14d1888f20d16e/src/python/pants/backend/python/rules/pytest_runner.py#L239

Finally, to actually publish the file into the `dist` dir at the end, you'd want to update https://github.com/pantsbuild/pants/blob/79b966ce77f746995e357ee06d14d1888f20d16e/src/python/pants/core/goals/test.py#L45 to include an optional test report XML name.

asherf replied:
One thing I can't figure out... is how to work w/Digest
How do I distinguish the coverage data from my data.
I see that coverage just consumes the entire digest:
https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/python/rules/pytest_runner.py#L248

Not sure how to handle multiple files/directories that are generated by the EPR..

Eric-Arellano commented on src/python/pants/core/goals/test.py:
```diff
@@ -185,6 +196,12 @@ def register_options(cls, register) -> None:
             "system supports this.",
         )
+        register(
+            "--xml-results",
```
I think we'll want to make this name more generic because this option impacts all test runners and not ever test runner uses XML. 

Tansy solved this for coverage by having the generic boolean option `--run-coverage`, then having Pytest Coverage have specific options for the report type and where to write the report in the subsystem `pytest-coverage`: 

https://github.com/pantsbuild/pants/blob/fca3558b3142acdecb5fe5730e6c5d342482e45b/src/python/pants/backend/python/rules/pytest_coverage.py#L218-L238

Here, the boolean option is right, but you'll want to rename the option to something more generic like `save-results` or `write-results`. Also update the `help` message.'

Then, you could add options to https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/python/subsystems/pytest.py to allow specifying where those results should be saved. 

If you want to support multiple formats for those results, you could make that an enum option like Tansy did with report types. If you only care about supporting XML for now, leave off that option for the type of results.

--

Next bit: update `pytest_runner.py` to a) propagate the relevant flags to Pytest, which you already do; then b) capture the relevant folder, which it looks like you also now do.

For splitting up the resulting digest into its relevant pieces, you'll use `SnapshotSubset`:

```python
coverage_digest = await Get[Digest](SnapshotSubset(output_digest, PathGlobs(".coverage"))
results_digest = await Get[Digest](SnapshotSubset(output_digest, PathGlobs(test_results_file))
```

--

Final bit: hook up `test.py` to write the reports to the build root, if any. This needs to happen in the `@goal_rule` because it's a side effect. See below for how we do it with coverage:

https://github.com/pantsbuild/pants/blob/fca3558b3142acdecb5fe5730e6c5d342482e45b/src/python/pants/core/goals/test.py#L291-L321

Your implementation can likely be substantially easier because a) we don't care about grouping related Results files, whereas Coverage needs this so that we can merge multiple individual results into one single report; and b), we don't have an option to open the results.

Your implementation will likely be closer to how we do `binary.py`:

https://github.com/pantsbuild/pants/blob/fca3558b3142acdecb5fe5730e6c5d342482e45b/src/python/pants/core/goals/binary.py#L114-L127

--

Finally, the tests will need some updating. I'm happy to pair on this or to send you some instructions when you're ready.

stuhood replied:
> I think we'll want to make this name more generic because this option impacts all test runners and not ever test runner uses XML.

junit XML is fairly standard at this point, and there are a lot of integrations that can consume it. It might be reasonable to expose it at the top level.

asherf commented:
Updated, also added junit_family as per https://docs.pytest.org/en/latest/deprecations.html#junit-family-default-value-change-to-xunit2
thanks to @Eric-Arellano for pairing with to help get this over the finish line!

Eric-Arellano commented on src/python/pants/backend/python/subsystems/pytest.py:
```diff
@@ -59,6 +59,22 @@ def register_options(cls, register):
+            type=str,
+            default="xunit2",
+            advanced=True,
+            help="JUnit XML format family",
```
It'd be helpful to link to the docs: https://docs.pytest.org/en/latest/reference.html#confval-junit_family

```suggestion
            help="The format of the generated XML file. See https://docs.pytest.org/en/latest/reference.html#confval-junit_family.",
```

Eric-Arellano commented on src/python/pants/backend/python/subsystems/pytest.py:
```diff
@@ -59,6 +59,22 @@ def register_options(cls, register):
+            metavar="<DIR>",
+            default=None,
+            advanced=True,
+            help="Specifying a directory causes junit xml results files to be emitted under "
```
Capitalization, since this will be seen by end users:

```suggestion
            help="Specifying a directory causes Junit XML result files to be emitted under "
```

Eric-Arellano commented on src/python/pants/backend/python/rules/pytest_runner.py:
```diff
@@ -228,26 +240,47 @@ async def run_python_test(
+    coverage_data = PytestCoverageData(config.address, output_digest) if run_coverage else None
+    xml_test_results_digest: Optional[Digest] = None
+    if test_setup.xml_dir:
+        snapshot = await Get[Snapshot](
```
It'd be good to call this `xml_snapshot` to make it more clear that this isn't as generic as `snapshot` implies.

Eric-Arellano commented on src/python/pants/backend/python/rules/pytest_runner.py:
```diff
@@ -228,26 +240,47 @@ async def run_python_test(
-        PytestCoverageData(config.address, result.output_directory_digest) if run_coverage else None
+    output_digest = result.output_directory_digest
+
+    coverage_data = PytestCoverageData(config.address, output_digest) if run_coverage else None
```
How come you got rid of the snippet to remove the XML results from `coverage_data` via `SnapshotSubset`? Does that still work properly? 

Note that we do not have any tests for coverage yet. Please run `./v2 --test-run-coverage test src/python/pants/util:tests` to ensure that it still works properly.

asherf replied:
@Eric-Arellano I don't follow... what do u think I removed?
I just extracted `result.output_directory_digest` into a local variable (`output_digest`) which makes the expression shorter so black consolidated it into a single line...
<img width="935" alt="Screen Shot 2020-04-27 at 6 24 02 PM" src="https://user-images.githubusercontent.com/1268088/80436556-474de700-88b4-11ea-89a1-3cdde3d0c327.png">

Eric-Arellano commented on src/python/pants/backend/python/rules/pytest_runner.py:
```diff
@@ -228,26 +240,47 @@ async def run_python_test(
     """Runs pytest for one target."""
-    env = {"PYTEST_ADDOPTS": f"--color={'yes' if global_options.options.colors else 'no'}"}
+    add_opts = [f"--color={'yes' if global_options.options.colors else 'no'}"]
+    test_results_file = f"{config.address.path_safe_spec}.xml"
```
This can be moved into the `if` statement because it's only needed when `xml_dir` is defined.

Eric-Arellano commented on src/python/pants/core/goals/test.py:
```diff
@@ -46,20 +46,25 @@ class TestResult:
     stderr: str
-    coverage_data: Optional["CoverageData"] = None
+    coverage_data: Optional["CoverageData"]
+    xml_test_results: Optional[Digest]
```
Maybe call it `xml_results`? I think the `test` part is obvious enough since the property lives on a `TestResult` class.

asherf commented:
<img width="1208" alt="Screen Shot 2020-04-27 at 6 28 22 PM" src="https://user-images.githubusercontent.com/1268088/80436815-e246c100-88b4-11ea-8261-246bdf42cb5d.png">
